doc/config.rst

```diff
@@ -70,6 +70,31 @@ and will first lookup global tox settings in this section:
     is identified. In a future version of tox, this warning will become an
     error.
 
+.. confval:: build=sdist|wheel|none
+
+    .. versionadded:: 3.1.0
+
+    tox will try to build the project as a Python package and install it in a fresh virtual
+    environment before running your test commands. Here one can select the type of build
+    tox will do for the projects package:
+
+    - if it's ``none`` tox will not try to build the package (implies no install happens later),
+    - if it's ``sdist`` it will create a source distribution by using ``distutils``:
+
+        .. code-block:: ini
+
+            python setup.py sdist
+
+    - if it's ``wheel`` it will create a Python wheel by using ``pip``:
+
+        .. code-block:: ini
+
+            pip build wheel --no-dep
```

Comment:
@asottile what you're describing there is what pip plans to do part of ``pip build wheel``.  I would prefer for us to not re-implement the build infrastructure and leave that to the build fronted, e.g. ``pip``.

Reply:
this would be a step backward until pip 10 has majority adoption and pep518 is actually implemented

Reply:
We don't have wheel support at the moment. So that's a step ahead. If we implement pep-518 in tox that would be totally redundant over implementing in pip. So I argue if we add this support let's do it in pip so everyone benefits. Inside tox you can upgrade pip, so I don't understand why you would need pip 10 majority adoption. You already get it with latest virtualenv which is installed even with pip 2.x automatically.

Reply:
The main point I'm making here is:

- tox is a tool to prevent incorrect packaging
- unless you're using pip 10 and pep518, building a wheel from the working directory can bypass some checks that tox currently enforces via the "make sdist then install from it" process (due to lack of workdir isolation)

imo we should either:
- require pip10 + pep518 support to enable `build=wheel`
- `sdist => wheel => install`

Here's an example project which tox would prevent today, but this PR would incorrecty allow with `build=wheel`:

### status quo, correctly failing

```
$ tail -n999 setup.py tox.ini requirements.txt 
==> setup.py <==
from setuptools import setup

# oops I goofed, should have made a MANIFEST.in
with open('requirements.txt') as f:
    contents = list(f)

setup(
    name='mylib',
    install_requires=contents,
)

==> tox.ini <==
[testenv]
commands =
    pip freeze --all

==> requirements.txt <==
six
```

```
$ tox -e py27
GLOB sdist-make: /private/tmp/g/setup.py
py27 create: /private/tmp/g/.tox/py27
py27 inst: /private/tmp/g/.tox/dist/mylib-0.0.0.zip
ERROR: invocation failed (exit code 1), logfile: /private/tmp/g/.tox/py27/log/py27-1.log
ERROR: actionid: py27
msg: installpkg
cmdargs: ['/private/tmp/g/.tox/py27/bin/pip', 'install', '/private/tmp/g/.tox/dist/mylib-0.0.0.zip']

Processing ./.tox/dist/mylib-0.0.0.zip
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/tmp/pip-req-build-CCk4wP/setup.py", line 3, in <module>
        with open('requirements.txt') as f:
    IOError: [Errno 2] No such file or directory: 'requirements.txt'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /private/tmp/pip-req-build-CCk4wP/

___________________________________ summary ____________________________________
ERROR:   py27: InvocationError for command /private/tmp/g/.tox/py27/bin/pip install /private/tmp/g/.tox/dist/mylib-0.0.0.zip (see /private/tmp/g/.tox/py27/log/py27-1.log) (exited with code 1)
```

### your branch, incorrectly succeeding

```ini
[tox]
build = wheel
[testenv]
commands =
    pip freeze --all
```

```
$ tox -e py36
GLOB wheel-make: /private/tmp/g/setup.py
py36 create: /private/tmp/g/.tox/py36
py36 inst: /private/tmp/g/.tox/dist/mylib-0.0.0-py3-none-any.whl
py36 installed: mylib==0.0.0,six==1.11.0
py36 runtests: PYTHONHASHSEED='1061640697'
py36 runtests: commands[0] | pip freeze --all
mylib==0.0.0
pip==10.0.1
setuptools==39.2.0
six==1.11.0
wheel==0.31.1
___________________________________ summary ____________________________________
  py36: commands succeeded
  congratulations :)
```

### your branch, unrelated also broken thing

(noticed this while making a reproduction) -- `tox`'s python is python3, but venvs is python2 so there's a wheel mismatch

(note that this is covered by my comment below about the open issue to use the _virtualenv_ python instead of `tox`'s python)

```
$ tox -e py27
GLOB wheel-make: /private/tmp/g/setup.py
py27 inst-nodeps: /private/tmp/g/.tox/dist/mylib-0.0.0-py3-none-any.whl
ERROR: invocation failed (exit code 1), logfile: /private/tmp/g/.tox/py27/log/py27-5.log
ERROR: actionid: py27
msg: installpkg
cmdargs: '/private/tmp/g/.tox/py27/bin/pip install -U --no-deps /private/tmp/g/.tox/dist/mylib-0.0.0-py3-none-any.whl'

mylib-0.0.0-py3-none-any.whl is not a supported wheel on this platform.

___________________________________ summary ____________________________________
ERROR:   py27: InvocationError for command /private/tmp/g/.tox/py27/bin/pip install -U --no-deps /private/tmp/g/.tox/dist/mylib-0.0.0-py3-none-any.whl (see /private/tmp/g/.tox/py27/log/py27-5.log) (exited with code 1)
```

Reply:
This branch wants to implement require pip10 + pep518 support to enable. Will check how come those succeed and alter the code/create pr for pip to fail. thanks!